Mathlib/Order/ReltoGaloisConnection.lean

```diff
@@ -0,0 +1,109 @@
+-/
+
+section
+
+universe u v
+
+variable {α : Type u} {β : Type v} (R : Rel α β)
+
+namespace Rel
+
+/-! ### Pairs of adjoint maps defined by relations -/
+
+open OrderDual
+
+/-- `leftDual` maps any set `J` of elements of type `α` to the set `{b : β | ∀ a ∈ J, R a b}` of
+elements `b` of type `β` such that `R a b` for every element `a` of `J`. -/
+def leftDual (J : Set α) : Set β := {b : β | ∀ ⦃a⦄, a ∈ J → R a b}
+
+/-- `rightDual` maps any set `I` of elements of type `β` to the set `{a : α | ∀ b ∈ ofDual I, R a b}`
+of elements `a` of type `α` such that `R a b` for every element `b` of `I`. -/
+def rightDual (I : Set β) : Set α := {a : α | ∀ ⦃b⦄, b ∈ I → R a b}
+
+/-- The pair of functions `leftDual` and `rightDual` forms a Galois connection. -/
+theorem gc_leftDual_rightDual : GaloisConnection (toDual ∘ R.leftDual) (R.rightDual ∘ ofDual) :=
+  fun J I ↦ ⟨fun h a ha b hb ↦ h (by simpa) ha, fun h b hb a ha ↦ h (by simpa) hb⟩
+
+/-! ### Induced equivalences and generation processes -/
+
+/-- `lFixedPoints` is the set of elements `J : Set α` satisfying `rightDual (leftDual J) = J`. -/
+def lFixedPoints := {J : Set α | R.rightDual (R.leftDual J) = J}
+
+/-- `rFixedPoints` is the set of elements `I : Set β` satisfying `leftDual (rightDual I) = I`. -/
+def rFixedPoints := {I : Set β | R.leftDual (R.rightDual I) = I}
+
+open GaloisConnection
+
+/-- `leftDual` maps every element `J` to `rFixedPoints`. -/
+theorem leftDual_mem_rFixedPoint (J : Set α) : R.leftDual J ∈ R.rFixedPoints := by
+  unfold rFixedPoints
+  apply le_antisymm
+  · apply R.gc_leftDual_rightDual.monotone_l; exact R.gc_leftDual_rightDual.le_u_l J
+  · exact R.gc_leftDual_rightDual.l_u_le (R.leftDual J)
+
+/-- `rightDual` maps every element `I` to `lFixedPoints`. -/
+theorem rightDual_mem_lFixedPoint (I : Set β) : R.rightDual I ∈ R.lFixedPoints := by
+  unfold lFixedPoints
+  apply le_antisymm
+  · apply R.gc_leftDual_rightDual.monotone_u; exact R.gc_leftDual_rightDual.l_u_le I
+  · exact R.gc_leftDual_rightDual.le_u_l (R.rightDual I)
+
+/-- The maps `leftDual` and `rightDual` induce inverse bijections between the sets of fixed points.
+-/
+def equivFixedPoints : R.lFixedPoints ≃ R.rFixedPoints where
+  toFun := fun ⟨J, _⟩ => ⟨R.leftDual J, R.leftDual_mem_rFixedPoint J⟩
+  invFun := fun ⟨I, _⟩ => ⟨R.rightDual I, R.rightDual_mem_lFixedPoint I⟩
+  left_inv J := by cases' J with J hJ; simp; rw [hJ]
+  right_inv I := by cases' I with I hI; simp; rw [hI]
+
+theorem u_l_le_of_le {J J' : Set α} (h : J' ∈ R.lFixedPoints) :
+    J ≤ J' → R.rightDual (R.leftDual J) ≤ J' := by
+  intro h₁
+  rw[← h]
+  apply R.gc_leftDual_rightDual.monotone_u
+  apply R.gc_leftDual_rightDual.monotone_l
+  exact h₁
+
+theorem ge_l_u_of_ge {I I' : Set β} (h : I' ∈ R.rFixedPoints) :
+    I' ≥ I → I' ≥ R.leftDual (R.rightDual I) := by
+  intro h₁
+  rw [← h]
+  apply R.gc_leftDual_rightDual.monotone_l
+  apply R.gc_leftDual_rightDual.monotone_u
+  exact h₁
```

Comment:
This can be turned into a proof term as follows:

```suggestion
theorem ge_l_u_of_ge {I I' : Set β} (h : I' ∈ R.rFixedPoints) :
    I' ≥ I → I' ≥ R.leftDual (R.rightDual I) :=
  fun h₁ ↦ h ▸ R.gc_leftDual_rightDual.monotone_l (R.gc_leftDual_rightDual.monotone_u h₁)
```

I have tested the code so you can safely commit this suggestion if you like.